fdbserver/RESTKmsConnector.actor.cpp

```diff
@@ -817,31 +827,30 @@ Future<T> kmsRequestImpl(
 						ctx->kmsUrlHeap.emplace(tempStack.top());
 						tempStack.pop();
 					}
-
 					return parsedResp;
 				} catch (Error& e) {
 					TraceEvent(SevWarn, "KmsRequestRespParseFailure").error(e).detail("RequestID", requestID);
 					curUrl->nResponseParseFailures++;
+					err = e;
 					// attempt to do request from next KmsUrl
 				}
 			} catch (Error& e) {
 				curUrl->nFailedResponses++;
-				if (pass > 1 && isKmsNotReachable(e.code())) {
-					TraceEvent(SevDebug, "KmsRequestFailedUnreachable", ctx->uid)
-					    .error(e)
-					    .detail("RequestID", requestID);
-					throw e;
-				} else {
-					TraceEvent(SevDebug, "KmsRequestError", ctx->uid).error(e).detail("RequestID", requestID);
-					// attempt to do request from next KmsUrl
-				}
+				err = e;
+				// attempt to do request from next KmsUrl
+				TraceEvent(SevDebug, "KmsRequestError", ctx->uid).error(e).detail("RequestID", requestID);
 			}
 		}
 
-		if (pass == 1) {
-			// Re-discover KMS urls and re-attempt request using newer KMS URLs
-			wait(discoverKmsUrls(ctx, RefreshPersistedUrls::True));
+		// If this is the first pass or we are getting connection errors then re-discover KMS urls and re-attempt
+		// request using newer KMS URLs
+		// TODO: Do we need exponential backoff here?
+		if (pass == 1 || (err.present() && isKmsNotReachable(err.get().code()))) {
+			wait(discoverKmsUrlsFn(ctx, RefreshPersistedUrls::True));
```

Comment:
Here if we time out/failed connection we rediscover kms urls from file. However if we have refresh the urls using previous requests then the file information might be stale which would always cause us to fail future requests. Not sure if this is the intended behaviour (this was the case even before these changes)? cc. @sfc-gh-ahusain